AdminServer/appscale/admin/instance_manager/source_manager.py

```diff
@@ -0,0 +1,146 @@
+""" Fetches and prepares the source code for revisions. """
+
+import logging
+import os
+import random
+import shutil
+
+from kazoo.exceptions import NodeExistsError
+from tornado import gen
+from tornado.options import options
+
+from appscale.common.appscale_utils import get_md5
+from .utils import fetch_file
+from ..constants import InvalidSource
+from ..constants import SOURCES_DIRECTORY
+from ..constants import UNPACK_ROOT
+from ..utils import extract_source
+
+
+class SourceManager(object):
+  """ Fetches and prepares the source code for revisions. """
+  def __init__(self, zk_client, thread_pool):
+    """ Creates a new SourceManager.
+
+    Args:
+      zk_client: A KazooClient.
+      thread_pool: A ThreadPoolExecutor.
+    """
+    self.zk_client = zk_client
+    self.thread_pool = thread_pool
+    self.source_futures = {}
+
+  @gen.coroutine
+  def fetch_archive(self, revision_key, source_location, try_existing=True):
+    """ Copies the source archive from a machine that has it.
+
+    Args:
+      revision_key: A string specifying a revision key.
+      source_location: A string specifying the location of the version's
+        source archive.
+      try_existing: A boolean specifying that the local file system should be
+        searched before copying from another machine.
+    Returns:
+      A string specifying the source archive's MD5 hex digest.
+    """
+    hosts_with_archive = yield self.thread_pool.submit(
+      self.zk_client.get_children, '/apps/{}'.format(revision_key))
+    assert hosts_with_archive, '{} has no hosters'.format(revision_key)
```

Comment:
I used it here just so I could raise an exception with a useful message. I could catch an IndexError during random.choice below, but I would need to define another custom exception type to raise since I couldn't find anything applicable.